### PR TITLE
tcp: extending tcp integration test

### PR DESCRIPTION
**Pull request by alyssawilk**
Adding some tests to ensure both the old and the new pool can handle many connections, and regression testing behavior as connections are added/used/removed over time.

Risk Level: n/a (test only)
Testing: yes
Docs Changes: no
Release Notes: no

### CODE REVIEW

File: test/integration/tcp_proxy_integration_test.cc [hunk 1]
```diff
@@ -150,6 +151,73 @@ TEST_P(TcpProxyIntegrationTest, TcpProxyDownstreamDisconnect) {
   tcp_client->waitForDisconnect();
 }
 
+TEST_P(TcpProxyIntegrationTest, TcpProxyManyConnections) {
+  autonomous_upstream_ = true;
+  initialize();
+  int num_connections = 50;
```
Comment (jmarantz) [inline test/integration/tcp_proxy_integration_test.cc:157]: nit: const

File: test/integration/tcp_proxy_integration_test.cc [hunk 2]
```diff
@@ -150,6 +151,73 @@ TEST_P(TcpProxyIntegrationTest, TcpProxyDownstreamDisconnect) {
...
+
+  for (int i = 0; i < 50; ++i) {
+    int action = rand.random() % 3;
+
+    if (action == 0) {
+      // Add a new connection.
+      clients.push_back(makeTcpConnection(lookupPort("tcp_proxy")));
+    }
+    if (clients.empty()) {
+      break;
```
Comment (jmarantz) [inline test/integration/tcp_proxy_integration_test.cc:193]: It looks like 2/3 of the time this test will do nothing; is that intended?

Do you want to seed the list with a few clients first?

File: test/integration/tcp_proxy_integration_test.cc [hunk 3]
```diff
@@ -150,6 +151,73 @@ TEST_P(TcpProxyIntegrationTest, TcpProxyDownstreamDisconnect) {
...
+  std::vector<IntegrationTcpClientPtr> clients(num_connections);
+
+  for (int i = 0; i < num_connections; ++i) {
+    clients[i] = makeTcpConnection(lookupPort("tcp_proxy"));
+  }
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_total", num_connections);
+  for (int i = 0; i < num_connections; ++i) {
+    IntegrationTcpClientPtr& tcp_client = clients[i];
+    ASSERT_TRUE(tcp_client->write(
+        "GET / HTTP/1.1\r\nHost: foo\r\nclose_after_response: yes\r\ncontent-length: 0\r\n\r\n",
```
Comment (jmarantz) [inline test/integration/tcp_proxy_integration_test.cc:167]: it seems unexpected that we are using HTTP manually in a TCP test. Do you want to comment on what the testing strategy is here?

**Review by jmarantz — COMMENTED** [review]
generally looks great and I think I understand what's going on, but can you add some motivation for this PR to the description?

**Comment by alyssawilk** [comment]
the motivation was that pinterest thought the new pool was having issues with many connections.  turns out it wasn't number of connections but throughput (#14453) but as long as I wrote the tests I think they're definitely a good one to have in our suite.

**Review by jmarantz — APPROVED** [review]

**Pull request merged**